### PR TITLE
 MAILBOX-372 InVM ErrorHandling for only group & use retry configuration in InVMEventBus

### PR DESCRIPTION
**Pull request by trantienduchn**
As I remember, Error handling is applied for only group registrations...

### CODE REVIEW

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java [hunk 1]
```diff
@@ -48,5 +48,5 @@
     }
 
 
-    ExecutionStages deliver(Collection<MailboxListener> mailboxListeners, Event event);
+    ExecutionStages deliver(Collection<InVMEventBus.TypedRegistration> typedRegistrations, Event event);
```
Comment (chibenwa) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]: Why does the InVmDelivery depends on an object specific to the InVmEventBus?

Why is the InVMEventBus taking the decision of retrying or not? I feels like the decision is hidded from an EventBus perspective.

My proposal, which might be simpler to implement is to have 2 methods : `deliver` && `deliverWithRetries`

**Comment by trantienduchn** [comment]
```
[c9a60e06559d14edc2d6dfc8e857a95dd18bb97a] [ERROR] Errors: 
[c9a60e06559d14edc2d6dfc8e857a95dd18bb97a] [ERROR]   CassandraRabbitMQJamesServerTest$MailsShouldBeWellReceived.mailsShouldBeWellReceived:70 » ConditionTimeout
[c9a60e06559d14edc2d6dfc8e857a95dd18bb97a] [ERROR]   CassandraRabbitMQJamesServerTest$MailsShouldBeWellReceived.mailsShouldBeWellReceived:70 » ConditionTimeout
[c9a60e06559d14edc2d6dfc8e857a95dd18bb97a] [ERROR]   CassandraRabbitMQJamesServerTest$MailsShouldBeWellReceived.mailsShouldBeWellReceived:70 » ConditionTimeout
[c9a60e06559d14edc2d6dfc8e857a95dd18bb97a] [INFO] 
[c9a60e06559d14edc2d6dfc8e857a95dd18bb97a] [ERROR] Tests run: 49, Failures: 0, Errors: 3, Skipped: 0
```
test this please

**Comment by trantienduchn** [comment]
Sorry there is another step to apply configuration POJO to InVMEventBus. I merge that step to this PR to avoid making many PRs

**Comment by chibenwa** [comment]
Can we rename the title of this PR to be matching what it is now doing?

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
```
[dd1a3037329762d632fd41ce0b4a54c28973dbb4] [ERROR] Errors: 
[dd1a3037329762d632fd41ce0b4a54c28973dbb4] [ERROR]   SpringMailboxTest.beforeClass:35 » BeanCreation Error creating bean with name ...
[dd1a3037329762d632fd41ce0b4a54c28973dbb4] [INFO] 
[dd1a3037329762d632fd41ce0b4a54c28973dbb4] [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0
```
test this please

**Comment by chibenwa** [comment]
```
SpringMailboxTest.beforeClass:35 » BeanCreation Error creating bean with name ...
```

It should be linked with changes of the constructor and is a real problem you need to solve.

**Comment by trantienduchn** [comment]
```
[0816d01035e8d9ddd73f4a67b0fce252e26cddbd] 1) org.apache.james.mailbox.events.delivery.InVmEventDelivery has more than one constructor annotated with @Inject. Classes must have either one (and only one) constructor annotated with @Inject or a zero-argument constructor that is not private.
[0816d01035e8d9ddd73f4a67b0fce252e26cddbd]   at org.apache.james.mailbox.events.delivery.InVmEventDelivery.class(InVmEventDelivery.java:49)
[0816d01035e8d9ddd73f4a67b0fce252e26cddbd]   at org.apache.james.modules.mailbox.DefaultEventModule.configure(DefaultEventModule.java:49) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.mailbox.CassandraMailboxModule -> org.apache.james.modules.mailbox.DefaultEventModule)
```
:roll_eyes:

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/InVmEventDelivery.java [hunk 2]
```diff
@@ -40,25 +41,44 @@
...
     private static final Logger LOGGER = LoggerFactory.getLogger(InVmEventDelivery.class);
-    private static final int MAX_RETRIES = 3;
-    private static final Duration FIRST_BACKOFF = Duration.ofMillis(100);
     private static final Duration MAX_BACKOFF = Duration.ofMillis(Long.MAX_VALUE);
-    private static final double DEFAULT_JITTER_FACTOR = 0.5;
 
     private final MetricFactory metricFactory;
+    private final RetryBackoffConfiguration retryBackoff;
 
-    @Inject
```
Comment (chibenwa) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/InVmEventDelivery.java]: Can we keep injecting this in Guice? + bind it there to the default instance.

You can call the other constructor in Spring by writing stuff in `mailbox/spring/src/main/resources/META-INF/spring/event-system.xml`

Comment (trantienduchn) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/InVmEventDelivery.java]: Spring works well with two `@Injected` constructors, only Guice has problem, it doesn't recognize good constructor to go.

Comment (chibenwa) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/InVmEventDelivery.java]: No you keep the @Inject on the one you want to use  with Guice as Spring config don't need @Inject

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/DefaultEventModule.java [hunk 3]
```diff
@@ -46,16 +49,20 @@ protected void configure() {
...
         bind(MailboxListenersLoader.class).to(MailboxListenersLoaderImpl.class);
-        bind(EventDelivery.class).to(InVmEventDelivery.class);
         bind(EventBus.class).to(InVMEventBus.class);
 
         Multibinder.newSetBinder(binder(), MailboxListener.GroupMailboxListener.class);
     }
 
+    @Provides
+    @Singleton
+    private EventDelivery eventDelivery(MetricFactory metricFactory) {
```
Comment (chibenwa) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/mailbox/DefaultEventModule.java]: No this can be done with a simple binding

**Comment by trantienduchn** [comment]
```
[6d5c273b77f933fc04acfef96a39a20b85a3b090] [ERROR] Errors: 
[6d5c273b77f933fc04acfef96a39a20b85a3b090] [ERROR]   JamesAppSpringMainTest.testServer:27 » BeanCreation Error creating bean with n...
[6d5c273b77f933fc04acfef96a39a20b85a3b090] [ERROR] org.apache.james.app.spring.JamesSpringContextTest.springShouldLoadAndAddOnlyOneQuotaUpdaterListener(org.apache.james.app.spring.JamesSpringContextTest)
[6d5c273b77f933fc04acfef96a39a20b85a3b090] [ERROR]   Run 1: JamesSpringContextTest.setup:45 » BeanCreation Error creating bean with name '...
[6d5c273b77f933fc04acfef96a39a20b85a3b090] [ERROR]   Run 2: JamesSpringContextTest.tearDown:52 NullPointer
```
:roll_eyes:

File: mailbox/spring/src/main/resources/META-INF/spring/event-system.xml [hunk 4]
```diff
@@ -20,14 +20,19 @@
 
 <beans xmlns="http://www.springframework.org/schema/beans"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns:util="http://www.springframework.org/schema/util"
        xsi:schemaLocation="
-          http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans.xsd">
+          http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans.xsd
```
Comment (trantienduchn) [inline mailbox/spring/src/main/resources/META-INF/spring/event-system.xml]: I followed [this answer](https://stackoverflow.com/a/11540129/6072746)

Comment (chibenwa) [reply mailbox/spring/src/main/resources/META-INF/spring/event-system.xml]: Also valid ^^

**Comment by trantienduchn** [comment]
```
[5d81cdff26e4e649ed93f11116b34d8230b540ca] [INFO] Apache James :: Server :: JMAP :: Cassandra Integration testing FAILURE [  02:00 h]
```
test this please

**Comment by trantienduchn** [comment]
```
[5d81cdff26e4e649ed93f11116b34d8230b540ca] [ERROR] Errors: 
[5d81cdff26e4e649ed93f11116b34d8230b540ca] [ERROR]   CassandraSetMailboxesMethodCucumberTest » ConditionTimeout Condition with lamb...
[5d81cdff26e4e649ed93f11116b34d8230b540ca] [INFO] 
[5d81cdff26e4e649ed93f11116b34d8230b540ca] [ERROR] Tests run: 94, Failures: 11, Errors: 1, Skipped: 0
```
probably my change set leads to a very bad integration test result

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
Warning: retrying makes spamassasin error host communication logs growing 3 times

**Comment by chibenwa** [comment]
> Warning: retrying makes spamassasin error host communication logs growing 3 times

Pb identified: https://github.com/linagora/james-project/pull/2111/commits/751777d4e4dbceb8b9c00aba59a9d19d77312250

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**